bin/check-kube-external-dns-services.rb

```diff
@@ -0,0 +1,135 @@
+         short: '-f FILTER',
+         long: '--filter'
+
+  option :exclude_namespace,
+         description: 'Exclude the specified list of namespaces',
+         short: '-n NAMESPACES',
+         long: '--exclude-namespace',
+         proc: proc { |a| a.split(',') },
+         default: ''
+
+  option :include_namespace,
+         description: 'Include the specified list of namespaces',
+         short: '-i NAMESPACES',
+         long: '--include-namespace',
+         proc: proc { |a| a.split(',') },
+         default: ''
+
+  HOSTNAME_ANNOTATION = 'external-dns.alpha.kubernetes.io/hostname'.freeze
+
+  # Returns a sorted array of external-dns hostnames set on the given service
+  def hostnames_for_service(service)
+    return [] if service.metadata.annotations.nil? || service.metadata.annotations[HOSTNAME_ANNOTATION].nil?
+    service.metadata.annotations[HOSTNAME_ANNOTATION].gsub(/\s/, '').split(',').sort
+  end
+
+  # Returns a sorted array of IPs for a given service that are exportable by external-dns
+  def addresses_for_service(service)
+    return [] if service.status.loadBalancer.nil? || service.status.loadBalancer.ingress.nil?
+    service.status.loadBalancer.ingress.map(&:ip).compact.sort
+  end
+
+  def should_exclude_namespace(namespace)
+    return !config[:include_namespace].include?(namespace) unless config[:include_namespace].empty?
+    config[:exclude_namespace].include?(namespace)
+  end
+
+  def run
+    bad_service_msg_map = {}
+
+    services = if config[:label_filter].nil?
+                 client.get_services(namespace: namespace)
+               else
+                 client.get_services(namespace: namespace, label_selector: config[:label_filter].to_s)
+               end
+
+    services.each do |s|
+      next if should_exclude_namespace(s.metadata.namespace)
+      next unless config[:services].empty? || config[:services].include?(s.metadata.name)
+
+      hostnames = hostnames_for_service(s)
+      next if hostnames.empty?
+
+      addresses = addresses_for_service(s)
+      next if addresses.empty?
+
+      msg_parts = []
+      hostnames.each do |hostname|
+        actual_addresses = Resolv.getaddresses(hostname).sort
+        next if addresses == actual_addresses
+
+        extra_ips = actual_addresses - addresses
+        missing_ips = addresses - actual_addresses
+
+        display_hostname = hostname.sub(/\.$/, '')
+        msg_parts << "#{display_hostname} has extra IPs #{extra_ips.inspect}" unless extra_ips.empty?
+        msg_parts << "#{display_hostname} is missing IPs #{missing_ips.inspect}" unless missing_ips.empty?
+      end
+
+      next if msg_parts.empty?
+      service_display_name = "#{s.metadata.namespace}/#{s.metadata.name}"
+      bad_service_msg_map.merge!(service_display_name => msg_parts.join(', '))
+    end
+
+    critical "The following failures were detected: #{bad_service_msg_map.inspect}" unless bad_service_msg_map.empty?
```

Comment:
This block is a little terse and hard to read.

Can we stick with the fairly conventional sensu patern of `if, elsif`? for readability and general consistency.

Reply:
pushed